Template part block: Add category panel

david-szabo97 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
Add category panel to template part block. This allows us to change the assigned area of the template part.

Related: https://github.com/WordPress/gutenberg/issues/29152
## How has this been tested?
1. Open site editor
2. Insert template part
3. Select template part block
4. Open settings sidebar
5. Switch to Block tab
6. You should see Category panel
7. Dropdown should show the correct value
8. Changing the value and saving should work

## Screenshots <!-- if applicable -->
![image](https://user-images.githubusercontent.com/2256104/108689230-20266d00-74f9-11eb-91e2-9169dde24129.png)


## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

david-szabo97 committed:
Add category panel

david-szabo97 commented on packages/block-library/src/template-part/edit/category-panel.js:
```diff
@@ -0,0 +1,34 @@
+		<PanelBody title={ __( 'Category' ) }>
+			<SelectControl
+				label={ __( 'Select template part category' ) }
+				labelPosition="top"
```
Any idea why this is not working? 🤔 The label is still positioned on the side.

ntsekouras replied:
I hope you don't mind a pushed a patch here: https://github.com/WordPress/gutenberg/pull/29159/commits/ffa154b2979fac2a3c1b9bb2c7c6e7fcb6d47b76

I'm not sure if this is the right way of handling this though, as it seems this PR: https://github.com/WordPress/gutenberg/pull/28609 introduced this 'conflict' (? 🤔 ). @sarayourfriend @ItsJonQ any feedback here?

david-szabo97 replied:
`Root` correctly applies the CSS based on the `labelPosition`:
![image](https://user-images.githubusercontent.com/2256104/108687741-5d89fb00-74f7-11eb-8906-ae89673c6145.png)

But for some reason a CSS class (which seems to be the `Flex` component) overrides it:
![image](https://user-images.githubusercontent.com/2256104/108687805-72ff2500-74f7-11eb-9296-154dec246054.png)


This is interesting because `Flex`'s styled CSS class should come BEFORE `Root`. `Root` extends `Flex`:
https://github.com/WordPress/gutenberg/blob/0105c9a2312272d768f1ecb77c0b9e6dca77d0a0/packages/components/src/input-control/styles/input-control-styles.js#L45-L52

Oh, and it works fine in Storybook. So it's something with the Editor itself?

sarayourfriend replied:
I think the issue is that we're using two different `styled` functions and the specificity of the Root selector is less than the base Flex selector. There's one that comes from `@wp-g2/styles` that I think using would fix this. @ItsJonQ what's the intended solution in this situation?

ItsJonQ replied:
As @sarayourfriend has pointed out, the issue is indeed caused by the 2 ways the CSS-in-JS styles are being compiled. Something we need to smooth out for the Component System integration.

> Do we actually need the extra specificity? I wonder what would happen if we did without it.

We do. Even before the next Component System (G2), we had to resort to hacks like `&&&` for certain components just so they would render correctly:

https://github.com/WordPress/gutenberg/blob/master/packages/components/src/input-control/styles/input-control-styles.js#L184

My intuition tells me that we need to adjust some props in the `input-control/input-base.js` to ensure it plays nicely with the new `Flex` component (rather than a CSS fix).

ItsJonQ replied:
Update! I have a working solution. It involves remapping props from `input-control` to feed into the new `flex` component. Will create a PR shortly

ItsJonQ replied:
Created a PR here :)
https://github.com/WordPress/gutenberg/pull/29226

ntsekouras committed:
patch label positioning

ntsekouras commented on packages/block-library/src/template-part/edit/category-panel.js:
```diff
@@ -0,0 +1,36 @@
+				value={ area }
+				direction="column"
+				justify="flex-start"
+				onChange={ ( value ) => setArea( value ) }
```
No need to add an extra callback here
```suggestion
				onChange={ setArea }
```

david-szabo97 replied:
Thanks!

ntsekouras commented on packages/block-library/src/template-part/edit/category-panel.js:
```diff
@@ -0,0 +1,36 @@
+	);
+
+	return (
+		<PanelBody title={ __( 'Category' ) }>
```
I'm not sure this should be in its own panel and with this title. Same remark about the below label `Select template part category`. I think we want to avoid `template part` wording and maybe just `Area` should be okay? 🤔 
@mtias

david-szabo97 replied:
For now, I followed @jameskoster 's mockup that you can find here: https://github.com/WordPress/gutenberg/issues/27875

ntsekouras reviewed (commented):
Thanks for working on this @david-szabo97 !

I left some comments but the main functionality was tested and works 👍

david-szabo97 committed:
Remove unnecessary callback

david-szabo97 committed:
Extract options

jameskoster commented:
Thinking things through a little more... I am not sure that the title + area options should be so prominent in this context. It could easily mislead users into believing that any changes they make here are local to the parent template. I know the multi-entity save flow clears that up, but that might be a bit late...

Therefore, I'm thinking it might be better to move these options to the "Advanced" section:

<img width="363" alt="Screenshot 2021-02-22 at 12 46 01" src="https://user-images.githubusercontent.com/846565/108710481-06425580-750c-11eb-99b2-a070661758c3.png">

The prominence could perhaps increase when editing a template part in isolation... something like:

<img width="1207" alt="Screenshot 2021-02-22 at 12 52 59" src="https://user-images.githubusercontent.com/846565/108711257-fd05b880-750c-11eb-995e-be47d75c12c4.png">

But that is something to explore in more detail elsewhere (#29148).

---

This is possibly tangential, but should the html element change when you switch area? IE if I switch from Header to Footer, should the tag also change from `<header>` to `<footer>`?

david-szabo97 commented:
> Therefore, I'm thinking it might be better to move these options to the "Advanced" section:

IMO, this makes sense. This is more like a power-user feature anyway.

> This is possibly tangential, but should the HTML element change when you switch area? IE if I switch from Header to Footer, should the tag also change from <header> to <footer>?

I'd say this is OK if the current area matches the current tag. IE, area = header, tag = header, but I switch to area = footer then tag becomes footer too. But if area = header, tag = div, then I switch to area = footer then tag should stay div.

Maybe that's too smart?

ntsekouras commented:
>I'd say this is OK if the current area matches the current tag. IE, area = header, tag = header, but I when switch to area = footer then tag becomes footer too. But if area = header, tag = div, then I switch to area = footer then tag should stay div.

> Maybe that's too smart?

I think it makes perfect sense to do exactly what you described.

mtias commented:
Agreed with placing them in "Advanced". I think "area" needs some contextual explanatory text. Ideally we could also use the icons on the select element.

Addison-Stavlo commented:
> I'd say this is OK if the current area matches the current tag. IE, area = header, tag = header, but I when switch to area = footer then tag becomes footer too. But if area = header, tag = div, then I switch to area = footer then tag should stay div.

I think it makes sense to only set a tag based on the semantic area if a tagName attribute isn't set.  So maybe we add an 'area default' to the tag name dropdown (and maybe add that default value in parens) that corresponds to the tagName being unset.  Then when no tagname is set, we default to that value that corresponds to the semantic area.

jameskoster commented:
We could use the existing icon buttons to create a simple segmented control:

<img width="283" alt="Screenshot 2021-02-22 at 15 55 06" src="https://user-images.githubusercontent.com/846565/108733393-60e8ab00-7526-11eb-9433-2dce688cfbe0.png">

Or we might create something more elaborate that can also be used in #29222 

<img width="283" alt="Screenshot 2021-02-22 at 15 55 56" src="https://user-images.githubusercontent.com/846565/108733511-78279880-7526-11eb-83c8-c37520b5dfa0.png">

The helper text needs optimisation.

Addison-Stavlo commented:
>We could use the existing icon buttons to create a simple segmented control

It looks nice!  Im not sure how many different 'areas' we plan to support in the future, but if we add a handful that second one with the large icons might become a bit weighty.

jameskoster commented:
It could probably scale in rows – like the Inserter. But yeah, the simpler implementation might be the way to go to begin with.

mtias commented:
The first one looks nice to me :)

carolinan commented:
I prefer it when the text with the area name visible, or even the select list. 
When the icons are used, I have to stop and _think,_ trying to interpret what they mean. The tooltip can't replace that.

david-szabo97 committed:
Move controls to advanced panel

david-szabo97 committed:
Move controls to Advanced panel

david-szabo97 committed:
Add default HTML tag that is based on the area

Addison-Stavlo commented:
This is working well!
I like the default element tag based on area for when no tagName is set!  However, it looks like as a result of that there is no way to set it as a `div` without the area being 'General'.  I wonder if we should just add `div` to the list for elements now so users can have the power to override the area default with something non-semantic if there is ever any need.

Addison-Stavlo commented:
> I prefer it when the text with the area name is visible, or even the select list.

Maybe it makes sense to start with the dropdown list in this PR.  And explore the icon designs in follow up?

david-szabo97 committed:
Add div as an option

david-szabo97 committed:
Fix tests

david-szabo97 commented:
* Moved controls to Advanced panel
* Added div as an option for HTML element
* Added a new "Default based on area" option for HTML elements
* Fixed tests

> Maybe it makes sense to start with the dropdown list in this PR. And explore the icon designs in follow up?

I agree, let's keep them separate. Easier to review and discuss as well.

Addison-Stavlo commented on packages/block-library/src/template-part/edit/index.js:
```diff
@@ -91,31 +93,12 @@ export default function TemplatePartEdit( {
 
 	return (
 		<>
-			<InspectorControls>
```
Nice call moving all this to another file.  Its definitely a bit more organized now.

Addison-Stavlo commented on packages/e2e-tests/specs/experiments/multi-entity-editing.test.js:
```diff
@@ -41,9 +41,12 @@ const createTemplatePart = async (
-	const nameInputSelector =
-		'.block-editor-block-inspector .components-text-control__input';
-	const nameInput = await page.waitForSelector( nameInputSelector );
+	const advancedPanelXPath = `//div[contains(@class,"interface-interface-skeleton__sidebar")]//button[@class="components-button components-panel__body-toggle"][contains(text(),"Advanced")]`;
+	const advancedPanel = await page.waitForXPath( advancedPanelXPath );
+	await advancedPanel.click();
```
I wonder if current e2e utils `ensureSidebarOpened`, `findSidebarPanelToggleButtonWithTitle`, `findSidebarPanelWithTitle`, may be useful here?

david-szabo97 replied:
Would be great, but just checked them and they are only working in `edit-post` context 😞

Addison-Stavlo replied:
> are only working in edit-post context

Isn't that the context this function is being run in?  It looks like it is used in test setup in before all, after `createNewPost` and before `siteEditor.visit()`.  Similarly it uses `openDocumentSettingsSidebar` a few lines above this which is also reliant on `edit-post*` selectors.

Addison-Stavlo reviewed (approved):
This looks great and tests well in both the editor and front end!

ockham commented on packages/block-library/src/template-part/edit/get-tag-based-on-area.js:
```diff
@@ -0,0 +1,9 @@
+const AREA_TAGS = {
+	footer: 'footer',
+	header: 'header',
+	unactegorized: 'div',
```
Typo :grimacing: 

```suggestion
	uncategorized: 'div',
```

ockham replied:
Or maybe we should consider renaming to something else, now that we're using "area" rather than "category.

Maybe `unassigned`? :thinking:

jameskoster replied:
Some template parts need not be assigned to a specific area, but "uncategorized" or "unassigned" might make them feel miss-placed or unorganised. That doesn't feel like the appropriate tone. 

Perhaps something like "General" or "Universal"?

Addison-Stavlo replied:
"uncategorized" or "unassigned" seems appropriate in the case that an item exists and this value has never been assigned or distinguished.  Or that no other existing value settings are appropriate.

"General" and "Universal" could definitely be 'areas', but they do not fit the same purpose as uncategorized/unassigned.  We cannot assume something is "Universal" just because it doesn't fit into the existing categories, and when an item is not universal and does not fit into existing categories something like unassigned/uncategorized serves a purpose.

Addison-Stavlo replied:
https://github.com/WordPress/gutenberg/pull/29937 to fix the typos.

> Perhaps something like "General" or "Universal"?

I just realized/remembered on the user facing side, 'uncategorized' is being displayed as "General".
